File: src/test/java/uk/gov/justice/hmpps/prison/service/curfews/CurfewStateTest.java
package uk.gov.justice.hmpps.prison.service.curfews;

import org.junit.jupiter.api.Test;
import uk.gov.justice.hmpps.prison.api.model.HomeDetentionCurfew;

import java.time.LocalDate;

import static org.assertj.core.api.Assertions.assertThat;

public class CurfewStateTest {
    @Test
    public void initialState() {
        assertThat(CurfewState.getState(
                HomeDetentionCurfew.
                        builder()
                        .build()))
                .isInstanceOf(InitialState.class);
    }

    @Test
    public void checksFailedState() {
        assertThat(CurfewState.getState(
                HomeDetentionCurfew.
                        builder()
                        .passed(Boolean.FALSE)
                        .checksPassedDate(LocalDate.now())
                        .build()
        )).isInstanceOf(ChecksFailedState.class);
    }

    @Test
    public void checksPassedState() {
        assertThat(CurfewState.getState(
                HomeDetentionCurfew.
                        builder()
                        .passed(Boolean.TRUE)
                        .checksPassedDate(LocalDate.now())
                        .build()
        )).isInstanceOf(ChecksPassedState.class);
    }

    @Test
    public void checksFailedRefusedState() {
        assertThat(CurfewState.getState(
                HomeDetentionCurfew.
                        builder()
                        .passed(Boolean.FALSE)
                        .checksPassedDate(LocalDate.now())
                        .approvalStatus("REJECTED")
                        .refusedReason("BREACH")
                        .approvalStatusDate(LocalDate.now())
                        .build()
        )).isInstanceOf(ChecksFailedRefusedState.class);
    }

    @Test
    public void checksPassedRefusedState() {
        assertThat(CurfewState.getState(
                HomeDetentionCurfew.
                        builder()
                        .passed(Boolean.TRUE)
                        .checksPassedDate(LocalDate.now())
                        .approvalStatus("REJECTED")
                        .refusedReason("BREACH")
                        .approvalStatusDate(LocalDate.now())
                        .build()
        )).isInstanceOf(ChecksPassedRefusedState.class);
    }

    @Test
    public void checksPassedApprovedState() {
        assertThat(CurfewState.getState(
                HomeDetentionCurfew.
                        builder()
                        .passed(Boolean.TRUE)
                        .checksPassedDate(LocalDate.now())
                        .approvalStatus("APPROVED")
                        .approvalStatusDate(LocalDate.now())
                        .build()
        )).isInstanceOf(ChecksPassedApprovedState.class);
    }
}

File: src/test/java/uk/gov/justice/hmpps/prison/service/curfews/CurfewStateTest.kt
package uk.gov.justice.hmpps.prison.service.curfews

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import uk.gov.justice.hmpps.prison.api.model.HomeDetentionCurfew
import java.time.LocalDate

class CurfewStateTest {
  @Test
  fun initialState() {
    assertThat(
      CurfewState.getState(
        HomeDetentionCurfew.builder
          ()
          .build(),
      ),
    )
      .isInstanceOf(InitialState::class.java)
  }

  @Test
  fun checksFailedState() {
    assertThat(
      CurfewState.getState(
        HomeDetentionCurfew.builder
          ()
          .passed(false)
          .checksPassedDate(LocalDate.now())
          .build(),
      ),
    ).isInstanceOf(ChecksFailedState::class.java)
  }

  @Test
  fun checksPassedState() {
    assertThat(
      CurfewState.getState(
        HomeDetentionCurfew.builder
          ()
          .passed(true)
          .checksPassedDate(LocalDate.now())
          .build(),
      ),
    ).isInstanceOf(ChecksPassedState::class.java)
  }

  @Test
  fun checksFailedRefusedState() {
    assertThat(
      CurfewState.getState(
        HomeDetentionCurfew.builder
          ()
          .passed(false)
          .checksPassedDate(LocalDate.now())
          .approvalStatus("REJECTED")
          .refusedReason("BREACH")
          .approvalStatusDate(LocalDate.now())
          .build(),
      ),
    ).isInstanceOf(ChecksFailedRefusedState::class.java)
  }

  @Test
  fun checksPassedRefusedState() {
    assertThat(
      CurfewState.getState(
        HomeDetentionCurfew.builder
          ()
          .passed(true)
          .checksPassedDate(LocalDate.now())
          .approvalStatus("REJECTED")
          .refusedReason("BREACH")
          .approvalStatusDate(LocalDate.now())
          .build(),
      ),
    ).isInstanceOf(ChecksPassedRefusedState::class.java)
  }

  @Test
  fun checksPassedApprovedState() {
    assertThat(
      CurfewState.getState(
        HomeDetentionCurfew.builder
          ()
          .passed(true)
          .checksPassedDate(LocalDate.now())
          .approvalStatus("APPROVED")
          .approvalStatusDate(LocalDate.now())
          .build(),
      ),
    ).isInstanceOf(ChecksPassedApprovedState::class.java)
  }
}
